refactor!: cleaning up our use of `testWallet.getNotes(...)`

benesjan opened the pull request:
In this PR I replace incorrect use of the `testWallet.getNotes(...)` debugging function by implementing corresponding utility functions and I drop the Aztec Cheatcodes class as it became unused.

Nuked the testing with Aztec.js section of docs as it became so stale to the point of being detrimental. Asked devrel and they prefer me rewriting it so I will do that in a followup PR.

benesjan commented:
* **#17380** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17380?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17325** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17325?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17325?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#17295** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17295?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17303** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17303?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17283** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17283?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/aztec-nr/aztec/src/note/note_metadata.nr:
```diff
@@ -40,7 +40,7 @@ global NoteStage: NoteStageEnum =
 /// order to access the underlying fields callers must first find the appropriate stage (e.g. via `is_settled()`) and
 /// then convert this into the appropriate type (e.g. via `to_settled()`).
-#[derive(Eq, Serialize, Packable)]
+#[derive(Deserialize, Eq, Serialize, Packable)]
```
We now return `RetrievedNote` from a contract function which forced me to do this change.

nventuro replied:
In general all of our `pub` structs should probably have these impls - if we expose them at all then that means that we return or take them, so people might end up doing this, but they won't be able to fix the problem on their end.

benesjan replied:
Will take a note to do this 👍

benesjan commented on noir-projects/aztec-nr/aztec/src/note/retrieved_note.nr:
```diff
@@ -1,9 +1,9 @@
 /// pending (created in the current transaction) or settled (created in a previous transaction).
-#[derive(Eq, Serialize, Packable)]
+#[derive(Deserialize, Eq, Serialize, Packable)]
 pub struct RetrievedNote<Note> {
```
We now return RetrievedNote from a contract function which forced me to do this change.

benesjan commented on noir-projects/noir-contracts/contracts/app/claim_contract/src/main.nr:
```diff
@@ -36,19 +36,20 @@ pub contract Claim {
         // 2) Prove that the note hash exists in the note hash tree
-        // Note: The note has been inserted into the donation_receipts set in the Crowdfunding contract.
-        let note_storage_slot = Crowdfunding::storage_layout().donation_receipts.slot;
+        // Note: The note has been inserted into the donation_receipts set in a map under the owner address in
+        // the Crowdfunding contract.
+        let note_storage_slot = derive_storage_slot_in_map(
+            Crowdfunding::storage_layout().donation_receipts.slot,
+            owner,
+        );
```
This change will be clear after looking at the Crowdfunding contract.

benesjan commented on noir-projects/noir-contracts/contracts/app/crowdfunding_contract/src/main.nr:
```diff
@@ -37,7 +42,7 @@ pub contract Crowdfunding {
         config: PublicImmutable<Config, Context>,
         // Notes emitted to donors when they donate (can be used as proof to obtain rewards, eg in Claim contracts)
-        donation_receipts: PrivateSet<UintNote, Context>,
+        donation_receipts: Map<AztecAddress, PrivateSet<UintNote, Context>, Context>,
```
Because of the `get_donation_notes` function I now need to have a proper `Map` here.

benesjan commented on noir-projects/noir-contracts/contracts/app/crowdfunding_contract/src/main.nr:
```diff
@@ -78,10 +83,12 @@ pub contract Crowdfunding {
             &mut context,
             donor,
-            MessageDelivery.CONSTRAINED_ONCHAIN,
+            MessageDelivery.UNCONSTRAINED_ONCHAIN,
```
Unrelated change. Using constrained encryption here was incorrect so I updated it.

benesjan commented on noir-projects/noir-contracts/contracts/app/crowdfunding_contract/src/main.nr:
```diff
@@ -110,4 +117,34 @@ pub contract Crowdfunding {
+
+        // We bypass Aztec.nr's higher-level abstractions here to access the metadata available only in RetrievedNote
+        // type that is not returned by view_notes function. We on the other hand need this metadata because we will
+        // be proving the note existence in the Claim contract.
```
This might be interesting for @nventuro 

Also, this is a great example where having a utility function be callable from another contract would be valuable (the Claim contract is the destination where this note ends up hence would make sense to just call it from there).

nventuro replied:
Right, instead of the TS code calling this and then passing that into the other contract, the claims contract could just call this utility fn as an oracle.

It is true however that it is strange that our `prove_inclusion` oracle requires a `RetrievedNote` but none of the functions return it, so we need to resort to the low level APIs to do it.

benesjan commented on yarn-project/pxe/src/pxe.ts:
```diff
@@ -1098,10 +1104,6 @@ export class PXE {
     return decodedEvents;
   }
 
-  async resetNoteSyncData() {
```
This was unused so I dropped it. It's only used by the internal PXE synchronizer so it shouldn't be exposed.

nventuro replied:
I think we may want to hold on to this until https://github.com/AztecProtocol/aztec-packages/issues/15753 is fixed (or more generally until tagging indeces become robust), since this is the only mechanism for wallets to fix PXE if the indeces get broken. @Thunkar this is what they're relying on no?

Thunkar replied:
I don't think anyone is using it and invoking the method won't get you out of #15753, because the problem is that your index "got outside" of the window. Resetting them won't unstuck you (nor cleaning PXEs db entirely)

benesjan replied:
In that case will not revert this change and all of Nico's feedback on this PR is now addressed (was careful with the commit history so seeing changes since last review should work).

benesjan commented on yarn-project/end-to-end/src/e2e_blacklist_token_contract/minting.test.ts:
```diff
@@ -85,21 +85,18 @@ describe('e2e_blacklist_token_contract mint', () => {
 
     describe('Mint flow', () => {
       it('mint_private as minter and redeem as recipient', async () => {
+        const balanceBefore = await asset.methods.balance_of_private(adminAddress).simulate({ from: adminAddress });
```
It was not justified to use the low level getNotes here so I replace with the balance_of_private call instead.

nventuro commented on noir-projects/aztec-nr/uint-note/src/uint_note.nr:
```diff
@@ -24,7 +24,9 @@ use dep::aztec::{
 /// A private note representing a numeric value associated to an account (e.g. a token balance).
-#[derive(Eq, Serialize, Packable)]
+/// Note: Deserialize trait implementation is derived here since in some contracts we are directly returning a UintNote
+/// from a contract function and macros require Deserialize to be derived when this is the case.
```
It'd remove this comment, a struct implemetning serde should not be a big deal. Also as mentioned above we probably want all pub structs to do it.

benesjan replied:
Dropped it in https://github.com/AztecProtocol/aztec-packages/pull/17325/commits/4b1c2cffb5cba82d7165b65c79453ca6a30355a4

nventuro reviewed (commented):
Lovely, thanks a lot for the careful considerations

benesjan committed:
refactor!: cleaning up our use of `testWallet.getNotes(...)`

In this PR I replace incorrect use of the `testWallet.getNotes(...)` debugging function by implementing corresponding utility functions and I drop the Aztec Cheatcodes class as it became unused.

Nuked the testing with Aztec.js section of docs as it became so stale to the point of being detrimental. Asked devrel and they prefer me rewriting it so I will do that in a followup PR.
